### PR TITLE
Tidy/fixes plus possible delegate

### PR DESCRIPTION
**Pull request by philsegeler**
Using a delegate class for event or tasks functions is not worth it, since
1. The overhead of ```std::function``` is negligible in this case
2. We lose the possibility to use ```std::bind``` and would force the user to use ```void*``` arguments.

However significant improvements were made in the performance and code cleanliness of the event handler, the task manager and ```shared_index_map```. Name2ID was fixed inside the container class to not be O(n) anymore.

Event and Regular task functions are now unified in the format:

```C++
enum class task_action { keep, discard };
typedef std::function<task_action(const oe::task_info_t)> method_type;
```

Test case is the ```events_task_test.cpp``` file where it simulates for 200 objects each collides with each, so 200*200 = 40000 events are created. 2000 more events are created as a buffer. Overall 42000 custom events.

However, the user will probably not handle all of them and not all of those colllisions happen each frame. So the test case simulates 10000 of those events called and about 1000 of them handled by the user each frame. If an event is not handled by the user it is still registered but no ```std::function``` is called. The functions supplied by the user are just a write to a map of strings.

The results are that the overhead of the 1000 user functions are about 45% of the CPU time of ```handleAllEvents```.

For 120 fps, the test consumes 70% single core CPU on the main thread and 20% on a second thread calling ```broadcast_event``` on my main pc. It utilizes 17-21% of the whole CPU on my older 2-core 4-thread laptop with 60 fps. It manages about 15-20 fps on the Intel Atom netbook.

The renderer also added another GL ES extension that I believed would increase the performance on the GMA 950, but it didn't... There is still no drawbacks of using it if it exists though.

Despite the above, there **still** exist possible use cases for delegate classes for the engine if we can use the delegates in an ```std::set``` or an ```std::map``` as comparison functions or in ```std::unordered_map``` and ```std::unordered_set``` as hash functions.

Such possibilities can be found in sorted ```shared_index_map.h``` and in ```types/polygon_storage.h```.

### CODE REVIEW

File: include/OE/error_oe.h
Comment (antsouchlos) [inline include/OE/error_oe.h:107]: Can't `data_` and `prefix_` be private? (The same goes for all exception classes)

File: include/OE/types/base_types.h
Comment (antsouchlos) [inline include/OE/types/base_types.h]: Can we easily rename this to `oe::name2id_container_t` or would that involve too many changes and be better suited for a later tidy branch?

File: include/OE/types/shared_index_map.h
Comment (antsouchlos) [inline include/OE/types/shared_index_map.h]: Can we rename this to `get_id()`?

File: include/OE/types/shared_index_map.h
Comment (antsouchlos) [inline include/OE/types/shared_index_map.h]: Can we rename this to `get_pointer()`?

File: include/OE/types/shared_index_map.h
Comment (antsouchlos) [inline include/OE/types/shared_index_map.h]: Rename to `iterator_t`?

File: include/OE/types/shared_index_map.h
Comment (antsouchlos) [inline include/OE/types/shared_index_map.h]: Should define the following:

```c++
using iterator_category = std::input_iterator_tag;
using difference_type   = std::ptrdiff_t;
using value_type        = element_t;
using reference         = value_type&;
using pointer           = value_type*;

const element_t* operator->() const {
    return &element_t(db_, (*iter).second);
}

const element_t& operator*() const {
    return element_t(db_, (*iter).second);
}
```

Source: [cppreference](https://en.cppreference.com/w/cpp/named_req/InputIterator)

File: include/OE/types/shared_index_map.h
Comment (antsouchlos) [inline include/OE/types/shared_index_map.h]: Same as for `Iterator` above

File: src/task.cpp
Comment (antsouchlos) [inline src/task.cpp]: Should this be `unsigned`?

**Comment by antsouchlos** [comment]
Btw: Jesus you madman, 70 files and `+2,936` `−2,163`. Nice code though

File: include/OE/types/shared_index_map.h
Comment (philsegeler) [reply include/OE/types/shared_index_map.h]: Defining a reference to a temporary gives me a warning and it makes sense. I will use ```const element_t``` instead and mark all public methods of ```element_t``` as ```const```

File: include/OE/types/base_types.h
Comment (philsegeler) [reply include/OE/types/base_types.h]: I made a remark that it is deprecated. It is only still used in ```types/vertex_storage.h``` and ```types/vertex_group.h```, because the way those are stored should be refactored and use a ```shared_index_map_t``` like everything else. Not yet now though.

File: include/OE/error_oe.h
Comment (philsegeler) [reply include/OE/error_oe.h:107]: Yes

File: include/OE/types/shared_index_map.h
Comment (philsegeler) [reply include/OE/types/shared_index_map.h]: Yes

File: include/OE/types/shared_index_map.h
Comment (philsegeler) [reply include/OE/types/shared_index_map.h]: Yes

File: include/OE/types/shared_index_map.h
Comment (philsegeler) [reply include/OE/types/shared_index_map.h]: Yes

File: include/OE/types/shared_index_map.h
Comment (philsegeler) [reply include/OE/types/shared_index_map.h]: Yes

File: src/task.cpp
Comment (philsegeler) [reply src/task.cpp]: Yes, you are right. I made them all ```std::size_t```.

**Comment by philsegeler** [comment]
Btw don't merge, yet I am trying to fix some other thing now regarding global variables

**Comment by philsegeler** [comment]
Ok now you can approve, if you like. I optimized the file includes a bit and now you do not have to recompile 260878 files if you change the lexer/parser/interpreter. :) Removed unnecessary ctre dependency from the executables.

I also put all ```shared_index_map_t``` global variables into their own separate file ```global_variables.h``` file as a temporary solution before a ```oe::world_manager_t``` class can be implemented as the one of the two main global variables of the engine.

**Review by antsouchlos — APPROVED** [review]

**Pull request merged**
